File: meldanya-python3/day5/day5.py
import sys


def jump(jumps, func):
    i = 0
    n = 0
    while i >= 0 and i < len(jumps):
        jump = jumps[i]
        jumps[i] = func(jumps[i])
        i += jump
        n += 1
    return n


def part1(jump):
    return jump + 1


def part2(jump):
    return jump - 1 if jump >= 3 else jump + 1


def main():
    jumps = [int(j) for j in sys.stdin]
    print(jump(jumps[:], part1))
    print(jump(jumps[:], part2))


if __name__ == '__main__':
    main()

File: digidis-go/day5/day5.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	part1()
	part2()
}

func part1() {
	f, _ := os.Open("input.txt")
	jumps := make([]int, 0)
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		i, _ := strconv.Atoi(scanner.Text())
		jumps = append(jumps, i)
	}
	f.Close()

	var pos, steps int
	for pos >= 0 && pos < len(jumps) {
		p := pos
		pos = pos + jumps[pos]
		jumps[p]++
		steps++
	}
	fmt.Printf("Answer: %v\n", steps)

}

func part2() {
	f, _ := os.Open("input.txt")
	jumps := make([]int, 0)
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		i, _ := strconv.Atoi(scanner.Text())
		jumps = append(jumps, i)
	}
	f.Close()

	var pos, steps int
	for pos >= 0 && pos < len(jumps) {
		p := pos
		pos = pos + jumps[pos]

		if jumps[p] >= 3 {
			jumps[p]--
		} else {
			jumps[p]++
		}

		steps++
	}
	fmt.Printf("Answer: %v\n", steps)

}
